File: frontend/src/components/benchmark/content/closeness_centrality/implementations/closeness_centrality.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class ClosenessCentrality
{
    private Dictionary<int, HashSet<int>> adjacencyList;

    public ClosenessCentrality()
    {
        adjacencyList = new Dictionary<int, HashSet<int>>();
    }

    public void AddEdge(int u, int v)
    {
        if (!adjacencyList.ContainsKey(u))
            adjacencyList[u] = new HashSet<int>();
        if (!adjacencyList.ContainsKey(v))
            adjacencyList[v] = new HashSet<int>();
        
        adjacencyList[u].Add(v);
        adjacencyList[v].Add(u);
    }

    public void BuildFromEdges(List<(int, int)> edges)
    {
        foreach (var (u, v) in edges)
        {
            AddEdge(u, v);
        }
    }

    public Dictionary<int, int> BfsDistances(int start)
    {
        var distances = new Dictionary<int, int>();
        var queue = new Queue<int>();
        
        distances[start] = 0;
        queue.Enqueue(start);
        
        while (queue.Count > 0)
        {
            int node = queue.Dequeue();
            foreach (int neighbor in adjacencyList.GetValueOrDefault(node, new HashSet<int>()))
            {
                if (!distances.ContainsKey(neighbor))
                {
                    distances[neighbor] = distances[node] + 1;
                    queue.Enqueue(neighbor);
                }
            }
        }
        
        return distances;
    }

    public double ClosenessCentrality(int node)
    {
        if (!adjacencyList.ContainsKey(node))
            return 0.0;
        
        var distances = BfsDistances(node);
        int reachableNodes = distances.Count - 1;
        
        if (reachableNodes == 0)
            return 0.0;
        
        int totalDistance = distances.Values.Sum();
        return (double)reachableNodes / totalDistance;
    }

    public List<(int, double)> AllClosenessCentralities()
    {
        var result = new List<(int, double)>();
        
        foreach (int node in adjacencyList.Keys)
        {
            result.Add((node, ClosenessCentrality(node)));
        }
        
        return result.OrderByDescending(x => x.Item2).ToList();
    }

    public List<(int, double)> TopKCentralNodes(int k = 10)
    {
        var centralities = AllClosenessCentralities();
        return centralities.Take(k).ToList();
    }

    public double NormalizedClosenessCentrality(int node)
    {
        if (!adjacencyList.ContainsKey(node))
            return 0.0;
        
        var distances = BfsDistances(node);
        int reachableNodes = distances.Count - 1;
        int totalNodes = adjacencyList.Count;
        
        if (reachableNodes == 0 || totalNodes <= 1)
            return 0.0;
        
        int totalDistance = distances.Values.Sum();
        return ((double)reachableNodes / totalDistance) * ((double)(totalNodes - 1) / reachableNodes);
    }
}

File: frontend/src/components/benchmark/content/closeness_centrality/implementations/closeness_centrality.rs
use std::collections::{HashMap, HashSet, VecDeque};

pub struct ClosenessCentrality {
    adjacency_list: HashMap<i32, HashSet<i32>>,
}

impl ClosenessCentrality {
    pub fn new() -> Self {
        ClosenessCentrality {
            adjacency_list: HashMap::new(),
        }
    }

    pub fn add_edge(&mut self, u: i32, v: i32) {
        self.adjacency_list.entry(u).or_insert_with(HashSet::new).insert(v);
        self.adjacency_list.entry(v).or_insert_with(HashSet::new).insert(u);
    }

    pub fn build_from_edges(&mut self, edges: &[(i32, i32)]) {
        for &(u, v) in edges {
            self.add_edge(u, v);
        }
    }

    pub fn bfs_distances(&self, start: i32) -> HashMap<i32, i32> {
        let mut distances = HashMap::new();
        let mut queue = VecDeque::new();
        
        distances.insert(start, 0);
        queue.push_back(start);
        
        while let Some(node) = queue.pop_front() {
            if let Some(neighbors) = self.adjacency_list.get(&node) {
                for &neighbor in neighbors {
                    if !distances.contains_key(&neighbor) {
                        distances.insert(neighbor, distances[&node] + 1);
                        queue.push_back(neighbor);
                    }
                }
            }
        }
        
        distances
    }

    pub fn closeness_centrality(&self, node: i32) -> f64 {
        if !self.adjacency_list.contains_key(&node) {
            return 0.0;
        }
        
        let distances = self.bfs_distances(node);
        let reachable_nodes = distances.len() - 1;
        
        if reachable_nodes == 0 {
            return 0.0;
        }
        
        let total_distance: i32 = distances.values().sum();
        reachable_nodes as f64 / total_distance as f64
    }

    pub fn all_closeness_centralities(&self) -> Vec<(i32, f64)> {
        let mut result = Vec::new();
        
        for &node in self.adjacency_list.keys() {
            result.push((node, self.closeness_centrality(node)));
        }
        
        result.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
        result
    }

    pub fn top_k_central_nodes(&self, k: usize) -> Vec<(i32, f64)> {
        let centralities = self.all_closeness_centralities();
        centralities.into_iter().take(k).collect()
    }

    pub fn normalized_closeness_centrality(&self, node: i32) -> f64 {
        if !self.adjacency_list.contains_key(&node) {
            return 0.0;
        }
        
        let distances = self.bfs_distances(node);
        let reachable_nodes = distances.len() - 1;
        let total_nodes = self.adjacency_list.len();
        
        if reachable_nodes == 0 || total_nodes <= 1 {
            return 0.0;
        }
        
        let total_distance: i32 = distances.values().sum();
        (reachable_nodes as f64 / total_distance as f64) * ((total_nodes - 1) as f64 / reachable_nodes as f64)
    }
}